### PR TITLE
Use rocky 9.4 release train snapshots for builds

### PR DESCRIPTION
**Pull request by wtripp180901**
Release of RockyLinux 9.5 has broken various things including OFED (currently OFED has no support for RL9.5). This PR fixes stackhpc image builds (only) by using the stackhpc ark repo snapshots for rocky and epel repos. After image builds the repos are left with the ark urls in, but disabled.

To support this, by default `ansible/site.yml` now runs with `appliances_mode: configure` which skips various non-idempotent tasks (e.g. `dnf install @groupname`.

**NB:** With this PR, only the `.stackhpc` environment will be able to build images (see pre/post hooks to enable/disable repos). This will be fixed in a later PR.

### CODE REVIEW

**Comment by wtripp180901** [comment]
TODO:
- nightly build
- fatimage
- ci

File: ansible/roles/release_train/templates/rocky-extras.repo.j2
[the remark this answers is not in the document]
Comment (assumptionsandg) [reply ansible/roles/release_train/templates/rocky-extras.repo.j2:1]: Would using the repository lists from SKC make sense here? https://github.com/stackhpc/stackhpc-kayobe-config/blob/ofed-fixes/etc/kayobe/dnf.yml

File: ansible/roles/dnf_repos/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  name: appstream
+  base_url: "{{ dnf_repos_rocky_ark_prefix }}/AppStream/{{ dnf_repos_rocky_ark_suffix }}"
+- file: rocky
+  name: crb
+  base_url: "{{ dnf_repos_rocky_ark_prefix }}/CRB/{{ dnf_repos_rocky_ark_suffix }}"
+- file: rocky-extras
+  name: extras
+  base_url: "{{ dnf_repos_rocky_ark_prefix }}/extras/{{ dnf_repos_rocky_ark_suffix }}"
+
+dnf_repos_epel_baseurl: https://ark.stackhpc.com/pulp/content/epel/9/Everything/x86_64/20240902T080424
```
Comment (assumptionsandg) [inline ansible/roles/dnf_repos/defaults/main.yml:24]: Should the timestamp here also be a variable? e.g. `dnf_repos_epel_ark_timestamp`

File: ansible/roles/dnf_repos/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+dnf_repos_rocky_ark_prefix: https://ark.stackhpc.com/pulp/content/{{ ansible_distribution | lower }}/9.4
+dnf_repos_rocky_ark_suffix: "{{ ansible_architecture }}/os/{{ dnf_repos_ark_timestamp }}/"
+# most stable from https://github.com/stackhpc/stackhpc-kayobe-config/blob/stackhpc/2024.1/etc/kayobe/pulp-repo-versions.yml
+# note that some timestamps can't be used because not all repos have snapshots for them
+dnf_repos_ark_timestamp: 20240816T002610
```
Comment (assumptionsandg) [inline ansible/roles/dnf_repos/defaults/main.yml:5]: nit: Would `dnf_repos_rocky_ark_timestamp` make more sense?

File: ansible/roles/release_train/templates/rocky-extras.repo.j2 [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+[extras]
+name=Rocky Linux $releasever - Extras
+baseurl={{ release_train_url_prefix }}/extras/{{ release_train_url_suffix }}
+gpgcheck=1
+enabled=1
+countme=1
+metadata_expire=6h
+gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-Rocky-9
+{{ release_train_auth }}
```
[the remark this answers is not in the document]
Comment (wtripp180901) [reply ansible/roles/release_train/templates/rocky-extras.repo.j2]: using these creds to attempt to basic auth with everything seems to cause issues when installing some repos later so probably best to leave them here for now

File: ansible/roles/dnf_repos/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+dnf_repos_rocky_ark_prefix: https://ark.stackhpc.com/pulp/content/{{ ansible_distribution | lower }}/9.4
```
Comment (assumptionsandg) [inline ansible/roles/dnf_repos/defaults/main.yml:1]: Could the Rocky version also be set to a fact here? e.g. `ansible_distribution_version`

**Comment by wtripp180901** [comment]
this should give an image which can run site.yml successfully, need to run it through the CI pipeline first if you're happy that its safe to do a nightly build (worked with packer locally)

File: environments/.stackhpc/hooks/pre.yml [hunk 5]
```diff
@@ -17,3 +17,12 @@
         - "{{ lookup('env', 'APPLIANCES_ENVIRONMENT_ROOT') }}/inventory/hosts.yml"
         - "{{ lookup('env', 'APPLIANCES_ENVIRONMENT_ROOT') }}/inventory/group_vars/all/secrets.yml"
         - "{{ lookup('env', 'APPLIANCES_ENVIRONMENT_ROOT') }}/inventory/group_vars/all/test_user.yml"
+
+- hosts: builder
+  become: yes
+  tasks:
+    - name: Replace system repos with ark
+      ansible.builtin.include_role:
+        name: release_train
```
[the remark this answers is not in the document]
Comment (wtripp180901) [reply environments/.stackhpc/hooks/pre.yml]: probably want to hold off moving this out of our CI into bootstrap until we can deploy site pulp servers

File: ansible/roles/openondemand/tasks/vnc_compute.yml [hunk 6]
```diff
@@ -48,6 +48,7 @@
   tags: install
   yum:
     name: '@Xfce'
+  when: appliances_mode != 'configure' # dnf group/module installs aren't idempotent so only run during build
```
Comment (sjpb) [inline ansible/roles/openondemand/tasks/vnc_compute.yml:51]: Ok - this task does run during fatimage.

File: packer/openhpc_extravars.yml [hunk 7]
```diff
@@ -1 +1,2 @@
 workaround_ansible_issue_61497: yes # extravars files can't be empty
+appliances_mode: "build"
```
Comment (sjpb) [inline packer/openhpc_extravars.yml:2]: Can we move this to environments/common/inventory/group_vars/builder/defaults.yml please? Just so there is only one place to look for things. And also remove the `"` just for consistency, barewords like this are fine.

**Review by sjpb — CHANGES_REQUESTED** [review]
Please address changes but I don't think there's a need to re-run CI.

File: ansible/roles/openondemand/tasks/vnc_compute.yml (hunk 6 above)
Comment (wtripp180901) [reply ansible/roles/openondemand/tasks/vnc_compute.yml:51]: yeah, I've tested a remote desktop when it's installed just in fatimage and it seems to work okay

File: environments/common/inventory/group_vars/all/openhpc.yml [hunk 8]
```diff
@@ -38,3 +38,13 @@ openhpc_config_default:
 openhpc_config_extra: {}
 openhpc_config: "{{ openhpc_config_default | combine(openhpc_config_extra, list_merge='append') }}"
 openhpc_state_save_location: "{{ appliances_state_dir + '/slurmctld' if appliances_state_dir is defined else '/var/spool' }}"
+
+ohpc_default_extra_repos:
```
[the remark this answers is not in the document]
Comment (wtripp180901) [reply environments/common/inventory/group_vars/all/openhpc.yml:42]: for now I guess, will need to remember to move it back out once we have client pulps though

**Review by sjpb — APPROVED** [review]
LGTM

**Pull request merged**